File: soma_maxima/soma_maxima_dividir_conquistar.cpp
/*
Soma máxima: dividir e conquistar
O(n log(n))
*/

#include <bits/stdc++.h>
using namespace std;

#define MAX 1000000
int a[MAX];

int divide_and_conquer(int left, int right)
{
    // condição de parada: um único elemento
    if (left == right)
    {
        // verifica se o valor é positivo
        if (a[left] > 0)
        {
            return a[left];
        }
        else
        {
            return 0;
        }
    }
    else
    {
    int sum = 0;
    int mid = (left + right) / 2;

    // soma parte esquerda
    int max_left = 0;
    for (int k = mid; k >= left; --k)
    {
        sum += a[k];
        if (sum > max_left)
        {
            max_left = sum;
        }
    }

    // soma parte direita
    int max_right = 0;
    sum = 0;
    for (int k = mid + 1; k <= right; ++k)
    {
        sum += a[k];
        if (sum > max_right)
        {
            max_right = sum;
        }
    }

    return max(
            max(
              divide_and_conquer(left, mid),
              divide_and_conquer(mid+1, right)
             ),
              max_left + max_right
          );
    }
}

int main()
{
    int n;

    while (scanf("%d", &n) != EOF)
    {
        for (int i = 0; i < n; ++i)
        {
            scanf("%d", &a[i]);
        }

        printf("%d\n", divide_and_conquer(0, n - 1));
    }

    return 0;
}

File: soma_maxima/soma_maxima_dividir_conquistar.py
def div_and_conquer(list):
	if len(list) == 1:
		if list[0] > 0:
			return list[0]
		else:
			return 0;
	else:
		sum = 0
		maxLeft = 0
		left = list[0:int(len(list)/2)]
		for x in reversed(range(0,len(left))):
			sum += left[x]

			if sum > maxLeft:
				maxLeft = sum

		sum = 0
		maxRight = 0
		right = list[int(len(list)/2):int(len(list))]
		for x in range(0,len(right)):
			sum += right[x]

			if sum > maxRight:
				maxRight = sum

		return max(max(div_and_conquer(left), div_and_conquer(right)), maxLeft+maxRight)

while True:
	try:
		n = int(input())
	except (EOFError):
		break

	inputList = []	
	for i in range(0,n):
		inputList.append(int(input()))
	
	print(div_and_conquer(inputList))

